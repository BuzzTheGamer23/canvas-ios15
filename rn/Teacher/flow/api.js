/* @flow */

export type ApiResponse<T> = AxiosResponse & {
  next: ?(() => Promise<ApiResponse<T>>),
}

export type ApiError = {
  response: {
    status: number,
    data: {
      errors: { message: string }[],
    },
  },
}
