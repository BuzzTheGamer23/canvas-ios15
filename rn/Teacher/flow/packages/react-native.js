// @flow
export type PanResponderGestureState = {
  stateID: number,
  moveX: number,
  moveY: number,
  x0: number,
  y0: number,
  dx: number,
  dy: number,
  vx: number,
  vy: number,
}
