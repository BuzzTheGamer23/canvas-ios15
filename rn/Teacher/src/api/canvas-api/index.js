// @flow

import * as coursesApi from './courses'
import * as usersApi from './users'
import * as enrollmentsApi from './enrollments'
import * as submissionsApi from './submissions'
import * as assignmentsApi from './assignments'
import * as quizzesApi from './quizzes'
import * as assignmentGroupsApi from './assignmentGroups'
import * as discussionsApi from './discussions'
import * as conversationApi from './conversations'

type CombinedApi = $Supertype<typeof coursesApi
  & typeof usersApi
  & typeof enrollmentsApi
  & typeof submissionsApi
  & typeof assignmentsApi
  & typeof assignmentGroupsApi
  & typeof quizzesApi
  & typeof discussionsApi
  & typeof conversationApi>

export default ({
  ...coursesApi,
  ...usersApi,
  ...enrollmentsApi,
  ...submissionsApi,
  ...assignmentsApi,
  ...quizzesApi,
  ...assignmentGroupsApi,
  ...discussionsApi,
  ...conversationApi,
}: CombinedApi)
