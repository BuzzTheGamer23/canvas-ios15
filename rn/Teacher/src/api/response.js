/* @flow */

export type ApiResponse<T> = {
  data: T,
  status: number,
  headers: {
    link: ?string,
  },
  next: ?(() => Promise<ApiResponse<T>>),
}

export type ApiError = {
  response: {
    status: number,
    data: {
      errors: { message: string }[],
    },
  },
}
