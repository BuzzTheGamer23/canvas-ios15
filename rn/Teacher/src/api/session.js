/* @flow */

export type Session = {
  authToken: string,
  baseURL: string,
}

let currentSession: ?Session = null

export function setSession (session: ?Session) {
  currentSession = session
}

export function getSession (): ?Session {
  return currentSession
}
