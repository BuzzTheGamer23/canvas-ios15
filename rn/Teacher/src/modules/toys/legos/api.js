/* this obviously won't be in the same module when we actually do API stuff */

export function buyLegoSet (legoSet: LegoSet): Promise<LegoSet> {
  return new Promise((resolve, reject) => {
    let action = () => {
      if (Math.random() < 0.3) {
        reject()
      } else {
        resolve(legoSet)
      }
    }
    setTimeout(action, 1000)
  })
}

export function sellAllLegos (): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, 1000))
}
