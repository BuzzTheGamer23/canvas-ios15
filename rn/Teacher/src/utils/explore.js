/* @flow */

function hasProp (object: any, key: string, value: any): boolean {
  if (object instanceof Object) {
    const props = object.props
    if (props && props instanceof Object) {
      for (var prop in props) {
        if (prop === key && props[prop] === value) {
          return true
        }
      }
    }
  }

  return false
}

function selectByProp (object: any, propKey: string, value: any): ?any {
  if (!(typeof object === 'object')) {
    return null
  }

  if (hasProp(object, propKey, value)) {
    return object
  }

  let result = null

  for (var key in object) {
    if (object[key] instanceof Array) {
      object[key].forEach((e) => {
        result = selectByProp(e, propKey, value)
        if (result) {
          return result
        }
      })
    }

    result = selectByProp(object[key], propKey, value)
    if (result) {
      return result
    }
  }

  return result
}

class ComponentExplorer {
  component: any

  constructor (component: any) {
    this.component = component
  }

  selectByProp (propKey: string, value: any): ?any {
    return selectByProp(this.component, propKey, value)
  }

  selectByID (id: string): ?any {
    return this.selectByProp('testID', id)
  }
}

export default function explore (component: any): ComponentExplorer {
  return new ComponentExplorer(component)
}
