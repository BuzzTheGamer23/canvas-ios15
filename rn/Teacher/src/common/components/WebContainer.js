//
// Copyright (C) 2016-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

/**
 * @flow
 */

import React, { Component } from 'react'
import { View, WebView } from 'react-native'
import RCTSFSafariViewController from 'react-native-sfsafariviewcontroller'
import { isWebUri } from 'valid-url'
import canvas, { getSession } from 'canvas-api'

const TEMPLATE = `
<html>
    <head>
        <meta name="viewport" content="width={{content-width}},
            initial-scale = 1.0, user-scalable = no" />
        <style>
            body {
                font: -apple-system-body;
                margin: 0;
                padding: 0;
                color: black;
                background-color: white;
            }
            img {
                width: auto;
                height: auto;
                max-width: 100%;
            }
            video {
                width: auto;
                height: auto;
                max-width: 100%;
            }
            #whizzy_content {
                padding: 0;
                margin: 0;
            }
        </style>
    </head>
<body>
    <div id='whizzy_content'>
    {{content}}
    </div>
    <script type="text/javascript">
    window.onload = function () {
        let interval = setInterval(function () {
            if (window.originalPostMessage) {
                let height = document.getElementById('whizzy_content').clientHeight;
                postMessage(JSON.stringify({type: 'UPDATE_HEIGHT', data: height }));
                clearInterval(interval)
            }
        }, 100)
    }
    </script>
</body>
</html>
`

type Props = {
  html: string,
  style?: any,
  scrollEnabled: boolean,
  contentInset?: { top?: number, left?: number, bottom?: number, right?: number },
  navigator?: Navigator,
}
export default class WebContainer extends Component<any, Props, any> {

  constructor (props: Props) {
    super(props)
    this.state = {
      webViewHeight: props.height || 1,
    }
  }

  onLayout = ({ nativeEvent }: { nativeEvent: { layout: { width: number }}}) => {
    const { width } = nativeEvent.layout
    if (width > 0) {
      this.setState({ viewportWidth: Math.ceil(width) })
    }
  }

  onShouldStartLoadWithRequest = (event: any): boolean => {
    if (event && event.navigationType === 'click' && event.url && isWebUri(event.url)) {
      const session = getSession()
      if (session && session.baseURL && event.url.includes(session.baseURL)) {
        this.loadAuthenticatedURL(event.url)
        return false
      }

      try {
        RCTSFSafariViewController.open(event.url)
      } catch (e) {}
      return false
    }
    return true
  }

  loadAuthenticatedURL = (url: string) => {
    let authedURL = url
    canvas.getAuthenticatedSessionURL(url).then(({ data }) => {
      if (data.session_url) {
        authedURL = data.session_url
      }
      RCTSFSafariViewController.open(authedURL)
    }).catch((e) => {
      RCTSFSafariViewController.open(authedURL)
    })
  }

  render () {
    let { html, style, scrollEnabled } = this.props
    scrollEnabled = scrollEnabled === undefined ? true : scrollEnabled

    html = TEMPLATE.replace('{{content}}', html)
    html = html.replace('{{content-width}}', `${this.state.viewportWidth}`)
    return (
      <View style={style} onLayout={this.onLayout} testID='web-container.view'>
        {
          this.state.viewportWidth == null
            ? null
            : <WebView
                style={{ height: this.state.webViewHeight, backgroundColor: 'rgba(0, 0, 0, 0)' }}
                source={{ html: html }}
                onMessage={this._onMessage}
                scrollEnabled={scrollEnabled}
                contentInset={this.props.contentInset}
                onShouldStartLoadWithRequest={this.onShouldStartLoadWithRequest}
            />
        }
      </View>
    )
  }

  _onMessage = (event) => {
    const message = JSON.parse(event.nativeEvent.data)
    switch (message.type) {
      case 'UPDATE_HEIGHT':
        this.setState({ webViewHeight: message.data })
        break
    }
  }
}
