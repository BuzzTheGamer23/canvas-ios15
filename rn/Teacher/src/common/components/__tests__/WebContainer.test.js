//
// Copyright (C) 2016-present Instructure, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

/**
 * @flow
 */

import 'react-native'
import React from 'react'
import WebContainer from '../WebContainer'
import renderer from 'react-test-renderer'
import explore from '../../../../test/helpers/explore'
import RCTSFSafariViewController from 'react-native-sfsafariviewcontroller'
import api, { setSession } from '../../../canvas-api'

jest
  .unmock('ScrollView')
  .mock('WebView', () => 'WebView')
  .mock('../../../routing/Screen')
  .mock('react-native-sfsafariviewcontroller', () => {
    return {
      open: jest.fn(),
    }
  })
  .mock('../../../canvas-api')

const template = {
  ...require('../../../__templates__/session'),
  ...require('../../../__templates__/helm'),
}

beforeAll(() => {
  setSession(template.session())
})

test('render', () => {
  let tree = renderer.create(
    <WebContainer />
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

test('render html', () => {
  let html = '<div>hello world</div>'
  let tree = renderer.create(
    <WebContainer html={html} />
  )
  var width = { nativeEvent: { layout: { width: 300 } } }
  tree.getInstance().onLayout(width)
  expect(tree.toJSON()).toMatchSnapshot()
})

test('render with width zero', () => {
  let html = '<div>hello world</div>'
  let tree = renderer.create(
    <WebContainer html={html} />
  )
  var width = { nativeEvent: { layout: { width: 0 } } }
  tree.getInstance().onLayout(width)
  expect(tree.toJSON()).toMatchSnapshot()
})

test('updates height from js', () => {
  let html = '<div>hello world</div>'
  let component = renderer.create(
    <WebContainer html={html} />
  )
  var width = { nativeEvent: { layout: { width: 300 } } }
  component.getInstance().onLayout(width)

  const webView: any = explore(component.toJSON()).query(({ type }) => type === 'WebView')[0]
  const data = JSON.stringify({ type: 'UPDATE_HEIGHT', data: 10 })
  const message = {
    nativeEvent: { data },
  }

  webView.props.onMessage(message)
  expect(component.toJSON()).toMatchSnapshot()
})

test('updates height from js with scroll disabled', () => {
  let html = '<div>hello world</div>'
  let component = renderer.create(
    <WebContainer html={html} scrollEnabled={false}/>
  )
  var width = { nativeEvent: { layout: { width: 300 } } }
  component.getInstance().onLayout(width)

  const webView: any = explore(component.toJSON()).query(({ type }) => type === 'WebView')[0]
  const data = JSON.stringify({ type: 'UPDATE_HEIGHT', data: 10 })
  const message = {
    nativeEvent: { data },
  }

  webView.props.onMessage(message)
  expect(component.toJSON()).toMatchSnapshot()
})

// External links
test('external links', () => {
  let html = '<div>hello world</div>'
  let tree = renderer.create(
    <WebContainer html={html} />
  )
  var width = { nativeEvent: { layout: { width: 300 } } }
  tree.getInstance().onLayout(width)
  const webView: any = explore(tree.toJSON()).query(({ type }) => type === 'WebView')[0]
  webView.props.onShouldStartLoadWithRequest()
  expect(RCTSFSafariViewController.open).not.toHaveBeenCalled()
  webView.props.onShouldStartLoadWithRequest({
    url: 'http://www.google.com',
    navigationType: 'click',
  })
  expect(RCTSFSafariViewController.open).toHaveBeenCalledWith('http://www.google.com')
})

test('external link that does not exist', () => {
  jest.resetAllMocks()
  let html = '<div>hello world</div>'
  let tree = renderer.create(
    <WebContainer html={html} />
  )
  var width = { nativeEvent: { layout: { width: 300 } } }
  tree.getInstance().onLayout(width)
  const webView: any = explore(tree.toJSON()).query(({ type }) => type === 'WebView')[0]
  webView.props.onShouldStartLoadWithRequest()
  expect(RCTSFSafariViewController.open).not.toHaveBeenCalled()
  webView.props.onShouldStartLoadWithRequest()
  expect(RCTSFSafariViewController.open).not.toHaveBeenCalled()
})

test('external link, then reload of the content', () => {
  jest.resetAllMocks()
  let html = '<div>hello world</div>'
  let tree = renderer.create(
    <WebContainer html={html} />
  )
  var width = { nativeEvent: { layout: { width: 300 } } }
  tree.getInstance().onLayout(width)
  const webView: any = explore(tree.toJSON()).query(({ type }) => type === 'WebView')[0]
  webView.props.onShouldStartLoadWithRequest()
  expect(RCTSFSafariViewController.open).not.toHaveBeenCalled()
  webView.props.onShouldStartLoadWithRequest({
    url: 'http://www.google.com',
    navigationType: 'click',
  })
  expect(RCTSFSafariViewController.open).toHaveBeenCalledWith('http://www.google.com')
  jest.resetAllMocks()
  webView.props.onShouldStartLoadWithRequest({
    url: 'about:blank',
    navigationType: 'other',
  })
  expect(RCTSFSafariViewController.open).not.toHaveBeenCalled()
})

test('internal link loads authenticated url', async () => {
  jest.resetAllMocks()
  let html = '<div>hello world</div>'
  const navigator = template.navigator({
    show: jest.fn(),
  })
  let tree = renderer.create(
    <WebContainer html={html} navigator={navigator} />
  )
  var width = { nativeEvent: { layout: { width: 300 } } }
  tree.getInstance().onLayout(width)
  const webView: any = explore(tree.toJSON()).query(({ type }) => type === 'WebView')[0]
  webView.props.onShouldStartLoadWithRequest()
  expect(RCTSFSafariViewController.open).not.toHaveBeenCalled()

  api.getAuthenticatedSessionURL.mockReturnValueOnce(Promise.resolve({ data: { session_url: 'http://mobiledev.instructure.com/courses/1/modules/1/items-authenticated' } }))

  await webView.props.onShouldStartLoadWithRequest({
    url: 'http://mobiledev.instructure.com/courses/1/modules/1/items',
    navigationType: 'click',
  })
  expect(RCTSFSafariViewController.open).toHaveBeenCalledWith('http://mobiledev.instructure.com/courses/1/modules/1/items-authenticated')
})
