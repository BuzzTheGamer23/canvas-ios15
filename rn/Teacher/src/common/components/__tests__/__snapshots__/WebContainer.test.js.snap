// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render 1`] = `
<View
  onLayout={[Function]}
  style={undefined}
  testID="web-container.view"
/>
`;

exports[`render html 1`] = `
<View
  onLayout={[Function]}
  style={undefined}
  testID="web-container.view"
>
  <WebView
    contentInset={undefined}
    onMessage={[Function]}
    onShouldStartLoadWithRequest={[Function]}
    scrollEnabled={true}
    source={
      Object {
        "html": "
      <html>
          <head>
              <meta name=\\"viewport\\" content=\\"width=300,
                  initial-scale = 1.0, user-scalable = no\\" />
              <style>
                  body {
                      font: -apple-system-body;
                      margin: 0;
                      padding: 0;
                      color: black;
                      background-color: white;
                  }
                  img {
                      width: auto;
                      height: auto;
                      max-width: 100%;
                  }
                  video {
                      width: auto;
                      height: auto;
                      max-width: 100%;
                  }
                  #whizzy_content {
                      padding: 0;
                      margin: 0;
                  }
              </style>
          </head>
      <body>
          <div id='whizzy_content'>
          <div>hello world</div>
          </div>
          <script type=\\"text/javascript\\">
          window.onload = function () {
              let interval = setInterval(function () {
                  if (window.originalPostMessage) {
                      let height = document.getElementById('whizzy_content').clientHeight;
                      postMessage(JSON.stringify({type: 'UPDATE_HEIGHT', data: height }));
                      clearInterval(interval)
                  }
              }, 100)
          }
          </script>
      </body>
      </html>
      ",
      }
    }
    style={
      Object {
        "backgroundColor": "rgba(0, 0, 0, 0)",
        "height": 1,
      }
    }
  />
</View>
`;

exports[`render with width zero 1`] = `
<View
  onLayout={[Function]}
  style={undefined}
  testID="web-container.view"
/>
`;

exports[`updates height from js 1`] = `
<View
  onLayout={[Function]}
  style={undefined}
  testID="web-container.view"
>
  <WebView
    contentInset={undefined}
    onMessage={[Function]}
    onShouldStartLoadWithRequest={[Function]}
    scrollEnabled={true}
    source={
      Object {
        "html": "
      <html>
          <head>
              <meta name=\\"viewport\\" content=\\"width=300,
                  initial-scale = 1.0, user-scalable = no\\" />
              <style>
                  body {
                      font: -apple-system-body;
                      margin: 0;
                      padding: 0;
                      color: black;
                      background-color: white;
                  }
                  img {
                      width: auto;
                      height: auto;
                      max-width: 100%;
                  }
                  video {
                      width: auto;
                      height: auto;
                      max-width: 100%;
                  }
                  #whizzy_content {
                      padding: 0;
                      margin: 0;
                  }
              </style>
          </head>
      <body>
          <div id='whizzy_content'>
          <div>hello world</div>
          </div>
          <script type=\\"text/javascript\\">
          window.onload = function () {
              let interval = setInterval(function () {
                  if (window.originalPostMessage) {
                      let height = document.getElementById('whizzy_content').clientHeight;
                      postMessage(JSON.stringify({type: 'UPDATE_HEIGHT', data: height }));
                      clearInterval(interval)
                  }
              }, 100)
          }
          </script>
      </body>
      </html>
      ",
      }
    }
    style={
      Object {
        "backgroundColor": "rgba(0, 0, 0, 0)",
        "height": 10,
      }
    }
  />
</View>
`;

exports[`updates height from js with scroll disabled 1`] = `
<View
  onLayout={[Function]}
  style={undefined}
  testID="web-container.view"
>
  <WebView
    contentInset={undefined}
    onMessage={[Function]}
    onShouldStartLoadWithRequest={[Function]}
    scrollEnabled={false}
    source={
      Object {
        "html": "
      <html>
          <head>
              <meta name=\\"viewport\\" content=\\"width=300,
                  initial-scale = 1.0, user-scalable = no\\" />
              <style>
                  body {
                      font: -apple-system-body;
                      margin: 0;
                      padding: 0;
                      color: black;
                      background-color: white;
                  }
                  img {
                      width: auto;
                      height: auto;
                      max-width: 100%;
                  }
                  video {
                      width: auto;
                      height: auto;
                      max-width: 100%;
                  }
                  #whizzy_content {
                      padding: 0;
                      margin: 0;
                  }
              </style>
          </head>
      <body>
          <div id='whizzy_content'>
          <div>hello world</div>
          </div>
          <script type=\\"text/javascript\\">
          window.onload = function () {
              let interval = setInterval(function () {
                  if (window.originalPostMessage) {
                      let height = document.getElementById('whizzy_content').clientHeight;
                      postMessage(JSON.stringify({type: 'UPDATE_HEIGHT', data: height }));
                      clearInterval(interval)
                  }
              }, 100)
          }
          </script>
      </body>
      </html>
      ",
      }
    }
    style={
      Object {
        "backgroundColor": "rgba(0, 0, 0, 0)",
        "height": 10,
      }
    }
  />
</View>
`;
